### PR TITLE
Speech translation

### PR DESCRIPTION
**Pull request by bearhsiang**
The results on COVOST2 en-de test set with default training hyper parameters in speech_translation/config.yaml are listed below. (SacreBleu)
- Fbank: 1.14
- APC: 4.15
- VQ-APC: 3.72
- NPC: 4.55
- Mockingjay: 3.82
- TERA: 4.13
- Modified CPC: 5.56
- VQ wav2vec: 6.39
- wav2vec2 15.03
- wav2vec2_base_960: 15.1
- wav2vec2_large_ll60k: 15.88
- hubert_base: 15.88
- hubert_large_ll60k: 19.89

README is in speech_translation/, please also help me put the text at proper place in downstream/README.md, thanks.

### CODE REVIEW

**Comment by leo19941227** [comment]
Hey @bearhsiang!

Thanks for the great work!
I have a quick feedback about the data preparation. Is it possible we provide a shell script for the `step1`?
I know that the steps are documented in the `covost` repo, but I think there are still lots of steps to follow.
So I am thinking that an automating script for data preparation will be nice.
What do you think? :)

**Comment by bearhsiang** [comment]
Ok, I will prepare the script to clone [covost2](https://github.com/facebookresearch/covost) and do the data splitting.
But voice data downloading still has to be done personally (some user information are needed)

**Comment by leo19941227** [comment]
Hey @bearhsiang!

Since I am a little bit overwhelmed recently, and I think for a benchmark it is important to have at least two people familiar with the entire task implementation to ensure the correctness, so it will need some time for another person to dive into the code and check the details (especially the evaluation metrics and the how we do testing.)
On the other hand, since @ftshijt is more experienced than me in ST. So I will let he help review this PR and discuss with you. Hopefully we can merge this PR in the following two weeks :)

Ps. @ftshijt will finish his intern this week and can work on this after then.

**Comment by bearhsiang** [comment]
Hi, @leo19941227 

That's great. Thanks!

**Comment by bearhsiang** [comment]
Hi, @leo19941227, @ftshijt

I have added the script about processing the covost2 dataset and data splitting.
Note that voice data still has to be downloaded personally.

File: s3prl/downstream/speech_translation/S3prl_SpeechToTextTask.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+    def __getitem__(
+        self, index: int
+    ) -> Tuple[int, torch.Tensor, Optional[torch.Tensor]]:
+
+        source, sr = torchaudio.load(self.audio_paths[index])
+
+        assert self.srs[index] == sr
+        source = self.resamplers[sr](source)
+        source = torch.mean(source, dim=0)
```
Comment (ftshijt) [inline s3prl/downstream/speech_translation/S3prl_SpeechToTextTask.py:180]: I'm not sure if this is OK for multi-channel data. In many cases, just use one channel would be fine. I'm not sure if the mean operation will have some side effects, let's check with others.

File: s3prl/downstream/speech_translation/S3prl_SpeechToTextTask.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+        source = source.view(-1)
+
+        if self.feature_transforms is not None:
+            assert not self.data_cfg.use_audio_input
+            source = self.feature_transforms(source)
+        source = source.float()
+
+        target = None
+        if self.tgt_texts is not None:
+            tokenized = self.tokenize_text(self.tgt_texts[index])
```
Comment (ftshijt) [inline s3prl/downstream/speech_translation/S3prl_SpeechToTextTask.py:190]: it's OK to do it on the fly, but how about making it pre-computed as you load the text? It will save some time from loaders, more or less.

File: s3prl/downstream/speech_translation/dataset.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+from torch.utils.data import Dataset
+from multiprocessing import Pool
+
+class DummyDataset(Dataset):
```
Comment (ftshijt) [inline s3prl/downstream/speech_translation/dataset.py:4]: I'm OK with the current dummy dataset settings, but would suggest a more elegant way :-)

File: s3prl/downstream/speech_translation/config.yaml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+  # configuration of decoding
+  generatorrc:
+    beam: 5
+    max_len_a: 0
+    max_len_b: 400
+
+  # the model configuration and creation are delegated to fairseq
+  modelrc:
+    arch: s2t_transformer
```
Comment (ftshijt) [inline s3prl/downstream/speech_translation/config.yaml:66]: Maybe I miss something: where I could find the detailed settings of the model (e.g., dropout, layer size, etc.)

File: s3prl/downstream/speech_translation/expert.py [hunk 5]
```diff
@@ -0,0 +1,506 @@
...
+            self.asr_bpe = encoders.build_bpe(Namespace(**asr_bperc))
+            
+            self.asr_decoder = TransformerDecoderScriptable(
+                self.model.decoder.args,
+                self.asr_dict,
+                Embedding(len(self.asr_dict), self.model.decoder.embed_dim, self.asr_dict.pad())
+            )
+            self.asr_task = S3prl_SpeechToTextTask.setup_task(Namespace(**downstream_expert['taskrc']))
+            self.asr_task.tgt_dict = self.asr_dict
+            self.asr_model = S2TTransformerModel(self.model.encoder, self.asr_decoder)
```
Comment (ftshijt) [inline s3prl/downstream/speech_translation/expert.py:110]: It's good to direct add an additional decoder for ASR. Another option would be to directly add a CTC layer to the encoder so that we won't be burdened too much for a large decoder. In espnet, we mostly do with the second way for efficient purpose.

File: s3prl/downstream/speech_translation/README.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
+## ST: Speech Translation
+
+specified by the command `-d speech_translation`
+
+#### Prepare data
```
Comment (ftshijt) [inline s3prl/downstream/speech_translation/README.md:5]: I notice many external python packages that are unique for ST. Maybe consider list a `requirement.txt` or present some environment guidance at the `readme`

**Review by ftshijt — COMMENTED** [review]
Great effort! I have some minor comments based on the codebase.

File: s3prl/downstream/speech_translation/config.yaml [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  datarc:
+    train: train
+    dev: dev
+    test: test
+    max_tokens: 10000
+    num_workers: 4
+
+  # configuration of decoding
+  generatorrc:
+    beam: 5
```
[the remark this answers is not in the document]
Comment (bearhsiang) [reply s3prl/downstream/speech_translation/config.yaml:60]: Yes, the beam size comes from [covost2 paper](https://arxiv.org/abs/2007.10310). I could test the results with different beam sizes and enlarge it if needed.
BTW, may I know which beam size you use in your experiment?

File: s3prl/downstream/speech_translation/expert.py [hunk 8]
```diff
@@ -0,0 +1,506 @@
...
+            with open(f'{self.output_prefix}-st-{mode}.tsv', 'w') as f:
+                print('id', 'hyp', 'ref', sep='\t', file=f)
+                for idx, hyp, ref in zip(records['ids'], records['hyps'], records['refs']):
+                    print(idx, hyp, ref, sep='\t', file=f)
+
+            for i in range(min(len(records['ids']), 5)):
+                tqdm.write(f"[st] [id]: {records['ids'][i]}, [hyp]: {records['hyps'][i]}, [ref]: {records['refs'][i]}")
+
+            tqdm.write(f'[BLEU]: {bleu.score}')
+            print(bleu)
```
[the remark this answers is not in the document]
Comment (bearhsiang) [reply s3prl/downstream/speech_translation/expert.py:480]: Yes, these lines are for debugging which show few samples for user to quickly check the decoding results and also the score. Since we already have output files and tensorboard to tract the result and performance, I think we could just remove them.

File: s3prl/downstream/speech_translation/README.md (hunk 6 above)
Comment (leo19941227) [reply s3prl/downstream/speech_translation/README.md:5]: Hi just for @bearhsiang 's information, we went from `requirement.txt` to `setup.py`, so you can update the package dependencies in `setup.py` at the root of the repo.

File: s3prl/downstream/speech_translation/S3prl_SpeechToTextTask.py (hunk 1 above)
Comment (bearhsiang) [reply s3prl/downstream/speech_translation/S3prl_SpeechToTextTask.py:180]: @ftshijt may I make sure that you means we could just select one channel (e.g. the first channel) and discard other channels?

BTW, @leo19941227 does s3prl support use multichannel waveform as input? if not, is there a consistence way to covert the multichannel data into mono in s3prl?

Comment (leo19941227) [reply s3prl/downstream/speech_translation/S3prl_SpeechToTextTask.py:180]: I think we will have a consistent way to do this, specifically for hidden set. I have tested mean & first channel on ASR and the difference is quite trivial on the hidden data. Actually I will preprocessed the audio consistently and distribute to each task in the single channel format. So we can first ignore this issue for now and the way you load waveform will be different eventually.

File: s3prl/downstream/speech_translation/S3prl_SpeechToTextTask.py (hunk 2 above)
Comment (bearhsiang) [reply s3prl/downstream/speech_translation/S3prl_SpeechToTextTask.py:190]: I think tokenizing on the fly is a more proper way. For the time consuming issue, the resampling and feature extraction would be the actual bottleneck. And if we have other hidden data to test on, it would be more convenient to take raw text as input.

Comment (ftshijt) [reply s3prl/downstream/speech_translation/S3prl_SpeechToTextTask.py:190]: Thanks for the clarification. Make sense to me

File: s3prl/downstream/speech_translation/config.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (ftshijt) [reply s3prl/downstream/speech_translation/config.yaml:60]: In my experiments, 10 to 30 seems to get a reasonable improvement over smaller beam size, then it will become not that significant.

File: s3prl/downstream/speech_translation/dataset.py (hunk 3 above)
Comment (bearhsiang) [reply s3prl/downstream/speech_translation/dataset.py:4]: The functionality of this DummyDataset class to help bridge the gap between Fairseq dataset and Torch dataset used in s3prl to support batch sampling with fixed tokens per batch, which could make the training more efficient.

There is another more elegant way which is modifying 
https://github.com/s3prl/s3prl/blob/205cdc3abd7d56169a98557bf865230a661a06de/s3prl/downstream/runner.py#L184-L193
By refreshing dataloader at the beginning of each epoch, we could directly using Fairseq dataset and remove DummyDataset.
Possible modification could be done as following codes
```python
# prepare data
dataloader = self.downstream.model.get_dataloader('train') 

batch_ids = []
backward_steps = 0
records = defaultdict(list)
epoch = self.init_ckpt.get('Epoch', 0)
while pbar.n < pbar.total:
   if is_initialized():
      # dataloader.sampler.set_epoch(epoch) 
      dataloader = self.downstream.refresh_dataloader(epoch, 'train') # <- add codes here
```
if @leo19941227 are interested in this modification, we could discuss more about it.

File: s3prl/downstream/speech_translation/config.yaml (hunk 4 above)
Comment (bearhsiang) [reply s3prl/downstream/speech_translation/config.yaml:66]: As I mention in [README](https://github.com/s3prl/s3prl/blob/ce4a33d101ac8d4887a1539ca8146433a196d790/s3prl/downstream/speech_translation/README.md/#training), I have delegate the model configuration and creation to Fairseq and the detailed settings are defined in https://github.com/pytorch/fairseq/blob/741fd138c69eebc1fac09b590047da85cdeafe2d/fairseq/models/speech_to_text/s2t_transformer.py

For example, the setting of architecture `s2t_transformer` is 
https://github.com/pytorch/fairseq/blob/741fd138c69eebc1fac09b590047da85cdeafe2d/fairseq/models/speech_to_text/s2t_transformer.py#L406-L446

If there are some confusion about the setting of hyper parameters, I could add more explanation in config.

File: s3prl/downstream/speech_translation/README.md (hunk 6 above)
Comment (bearhsiang) [reply s3prl/downstream/speech_translation/README.md:5]: Ok! I will add the dependence there.

**Comment by leo19941227** [comment]
@bearhsiang !

I just think of a point I am not sure whether you have already done it.
We should dump prediction files during evaluation, aka dev and test.
This will be good for further analysis and submission to leaderboard.
Also, we should provide a way to compute BLEU offline.
That is, a script giving BLEU depending on the prediction files and groundtruth file.
Many thanks ><

File: s3prl/downstream/speech_translation/config.yaml (hunk 4 above)
Comment (ftshijt) [reply s3prl/downstream/speech_translation/config.yaml:66]: Many thanks for the explanation. Then just for demonstration purposes, it would be great to add more explanation (probably also how to change that in the config)

**Comment by bearhsiang** [comment]
@leo19941227 I am wondering that is there any format about the prediction files, or I could just defined it by myself.

File: s3prl/downstream/speech_translation/expert.py (hunk 5 above)
Comment (bearhsiang) [reply s3prl/downstream/speech_translation/expert.py:110]: I think adding a CTC layer is a better way too and I will change the setting.

**Comment by leo19941227** [comment]
> @leo19941227 I am wondering that is there any format about the prediction files, or I could just defined it by myself.

幹抱歉，我直接漏回這邊了，如果發現我太久沒回請不要吝嗇直接私訊我XD
我覺得可以看 ST 有沒有什麼最常見的格式？想請問 @ftshijt 有沒有什麼建議呢？
不過如果沒有的話也是可以我們自己訂就好。在輸出的時候會需要 predicted 和 grountruth 都輸出到 expdir 哦。

**Comment by ftshijt** [comment]
> > @leo19941227 I am wondering that is there any format about the prediction files, or I could just defined it by myself.
> 
> 幹抱歉，我直接漏回這邊了，如果發現我太久沒回請不要吝嗇直接私訊我XD
> 我覺得可以看 ST 有沒有什麼最常見的格式？想請問 @ftshijt 有沒有什麼建議呢？
> 不過如果沒有的話也是可以我們自己訂就好。在輸出的時候會需要 predicted 和 grountruth 都輸出到 expdir 哦。

Cool! I think you can define the format on your own. We often use a very simple format, e.g., "<utt_id> <prediction>" per line.

File: s3prl/downstream/speech_translation/dataset.py (hunk 3 above)
Comment (leo19941227) [reply s3prl/downstream/speech_translation/dataset.py:3]: Hey @bearhsiang, yes! Actually I am planning to update the runner to let it re-get the dataloader from the downstream expert after each epoch. Also the `get_dataloader` function will have a new argument: `epoch: int`. Do you think this proposal can help you to make this part of code more elegant?

Comment (leo19941227) [reply s3prl/downstream/speech_translation/dataset.py:3]: Just for a comparison. Currently our pipeline get the dataloader at the start for the entire training, and then iterate over the dataloader again and again, where the dataloader and the dataset inside it can't adjust according to the epoch.

Comment (bearhsiang) [reply s3prl/downstream/speech_translation/dataset.py:3]: @leo19941227 

Yes, that would be great, and we could avoid using DummyDataset in this case.

Comment (leo19941227) [reply s3prl/downstream/speech_translation/dataset.py:3]: The get_dataloader with the epoch is supported at #203
